### DISCUSSION TITLE
Allow entities to change device

### DISCUSSION THREAD

**Discussion by kamaradclimber** [home-assistant/architecture#1055] [Entity Models] [answered]
Current implementation of HA requests entity `device_info` when the entity is added to HA ([source](https://github.com/home-assistant/core/blob/e91a38eede5131aebe1ed6d03ce9debe1994ab6a/homeassistant/helpers/entity_platform.py#L736)).

This works is most cases because link between an entity and its device is rather static (it does not change over time) and usually easy to know (for instance a temperature entity belong to a thermometer).

However, for some cases, the device might not be known in advance.
This is the case for templates since the list of entities used by a template is not known until the template is executed for the first time (and it can even change over time).

See https://github.com/home-assistant/architecture/discussions/1054 for the full context.

**Comment by emontnemery** [comment, marked as answer]
@kamaradclimber this is already supported by the existing APIs. What you link to is only there to make it convenient for developers to automatically create and update device registry entries, but it's not mandatory to use the entity's `device_info` property, and not all integrations do.

Usually, integrations directly call the device registry API to create devices without entities, but it would also be possible to do what you propose.
